src/libsyntax/ext/expand.rs

```diff
@@ -531,11 +531,34 @@ impl<'a, 'b> MacroExpander<'a, 'b> {
         let path = &mac.node.path;
 
         let ident = ident.unwrap_or_else(|| keywords::Invalid.ident());
-        let validate_and_set_expn_info = |def_site_span,
+        let validate_and_set_expn_info = |self_: &mut Self, // arg instead of capture
+                                          def_site_span,
                                           allow_internal_unstable,
-                                          allow_internal_unsafe| {
+                                          allow_internal_unsafe,
+                                          // can't infer this type
+                                          unstable_feature: Option<(Symbol, u32)>| {
+
+            // feature-gate the macro invocation
+            if let Some((feature, issue)) = unstable_feature {
+                // only if the outer span doesn't allow unstable invocations
+                // TODO: compare crates of span and def_site_span (can't figure out how)
```

Comment:
We only need to check stability for external crates but I can't figure out how to get the crate ID of a span or macro def in this context.